src/partition/ReceivedPartition.cpp

```diff
@@ -470,143 +488,312 @@ void ReceivedPartition::createOwnerInformation()
+        int counter = 0;
+        for (size_t i = 0; i < slaveOwnerVecs[rank].size(); i++) {
+          // Vertex has no owner yet and rank could be owner
+          if (globalOwnerVec[slaveGlobalIDs[rank][i]] == 0 && slaveTags[rank][i] == 1) {
+            slaveOwnerVecs[rank][i]                 = 1; // Now rank is owner
+            globalOwnerVec[slaveGlobalIDs[rank][i]] = 1; // Vertex now has owner
+            counter++;
+            if (counter == localGuess)
+              break;
+          }
+        }
+      }
+
+      // Second round: distribute all other vertices in a greedy way
+      PRECICE_DEBUG("Decide owners, second round in greedy way");
+      for (int rank = 0; rank < utils::MasterSlave::getSize(); rank++) {
+        for (size_t i = 0; i < slaveOwnerVecs[rank].size(); i++) {
+          if (globalOwnerVec[slaveGlobalIDs[rank][i]] == 0 && slaveTags[rank][i] == 1) {
+            slaveOwnerVecs[rank][i]                 = 1;
+            globalOwnerVec[slaveGlobalIDs[rank][i]] = rank + 1;
+          }
+        }
+      }
+
+      // Send information back to slaves
+      for (int rank = 1; rank < utils::MasterSlave::getSize(); rank++) {
+        if (not slaveTags[rank].empty())
+          PRECICE_DEBUG("Send owner information to slave rank " << rank);
+        utils::MasterSlave::_communication->send(slaveOwnerVecs[rank], rank);
+      }
+      // Master data
+      PRECICE_DEBUG("My owner information: " << slaveOwnerVecs[0]);
+      setOwnerInformation(slaveOwnerVecs[0]);
 
 #ifndef NDEBUG
-    for (size_t i = 0; i < globalOwnerVec.size(); i++) {
-      if (globalOwnerVec[i] == 0) {
-        PRECICE_DEBUG("The Vertex with global index " << i << " of mesh: " << _mesh->getName()
-                                                      << " was completely filtered out, since it has no influence on any mapping.");
+      for (size_t i = 0; i < globalOwnerVec.size(); i++) {
+        if (globalOwnerVec[i] == 0) {
+          PRECICE_DEBUG("The Vertex with global index " << i << " of mesh: " << _mesh->getName()
+                        << " was completely filtered out, since it has no influence on any mapping.");
+        }
       }
-    }
 #endif
-    auto filteredVertices = std::count(globalOwnerVec.begin(), globalOwnerVec.end(), 0);
-    if (filteredVertices)
-      PRECICE_WARN(filteredVertices << " of " << _mesh->getGlobalNumberOfVertices()
-                                    << " vertices of mesh " << _mesh->getName() << " have been filtered out "
-                                    << "since they have no influence on the mapping.");
+      auto filteredVertices = std::count(globalOwnerVec.begin(), globalOwnerVec.end(), 0);
+      if (filteredVertices)
+        PRECICE_WARN(filteredVertices << " of " << _mesh->getGlobalNumberOfVertices()
+                     << " vertices of mesh " << _mesh->getName() << " have been filtered out "
+                     << "since they have no influence on the mapping.");
+    }
```

Comment:
I guess those could be helpful for the two-level variant as well.

Reply:
Agreed and added.